File: python/algo/202507/algo0437.py
from collections import defaultdict
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 二叉树 - 2.7 回溯
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        cnt = defaultdict(int)
        cnt[0] = 1
        
        def dfs(node: Optional[TreeNode], sum: int) -> int:
            if node is None: return 0
            sum += node.val
            res = cnt[sum-targetSum]
            cnt[sum] += 1
            res += dfs(node.left, sum)
            res += dfs(node.right, sum)
            cnt[sum] -= 1   # 恢复现场
            return res
        
        return dfs(root, 0)
    

File: cplus/leetcode/202507/algo0437.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// 二叉树 - 2.7 回溯
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        // 自行解答
        unordered_map<long, int> cnt;

        auto dfs = [&](this auto&& dfs, TreeNode* node, long sum) -> int {
            if (!node) return 0;
            sum += node->val;
            auto res = cnt[sum-targetSum];
            cnt[sum] ++;
            res += dfs(node->left, sum); res += dfs(node->right, sum);
            cnt[sum] --;
            return res;
        };

        cnt[0] = 1;
        return dfs(root, 0);  
        
    }
};